apps/oracles/eth-rpc/src/lib.rs

```diff
@@ -144,128 +144,198 @@ type MyProvider = alloy::providers::fillers::FillProvider<
-        let provider = ProviderBuilder::new().connect_http(rpc_url.clone());
-        let eth_call = match feed_config.feed_id {
-            100001 => yn_ethx_call(provider.clone(), 12345 + 998),
-            100002 => yield_fi_yusd_call(provider.clone(), 12345 + 128),
-            100003 => yn_bnbx_call(provider.clone(), 12345 + 129),
-            _ => {
-                break;
+impl Contract {
+    pub fn prepare_request(&self, provider: MyProvider, id: u64) -> Option<RequestEthCall> {
+        match self.method_name.as_str() {
+            "convertToAssets" => Some(self.convert_to_assets(provider.clone(), id)),
+            "exchangeRate" => Some(self.exchange_rate(provider.clone(), id)),
+            _ => None,
+        }
+    }
+
+    pub fn convert_to_assets(&self, provider: MyProvider, id: u64) -> RequestEthCall {
+        let contact = VaultABI::new(self.address, provider.clone());
+        let shares = self.param1.unwrap();
+        let x = contact.convertToAssets(shares);
+        let calldata = x.calldata().clone();
+        RequestEthCall::latest(&calldata, &self.address, id)
+    }
+
+    pub fn exchange_rate(&self, provider: MyProvider, id: u64) -> RequestEthCall {
+        let contact = YieldFiyUSD::new(self.address, provider.clone());
+        let x = contact.exchangeRate();
+        let calldata = x.calldata().clone();
+        RequestEthCall::latest(&calldata, &self.address, id)
+    }
+
+    async fn fetch(&self, id: u64) -> Option<ResponseEthCall> {
+        let mut last_value = None;
+        for rpc_url_candidate in &self.rpc_urls {
+            if let Ok(rpc_url) = rpc_url_candidate.as_str().parse::<Url>() {
+                let provider = ProviderBuilder::new().connect_http(rpc_url.clone());
+                let eth_call = match self.prepare_request(provider, id) {
+                    Some(value) => value,
+                    None => break,
+                };
+                if let Ok(mut value) =
+                    http_post_json::<RequestEthCall, ResponseEthCall>(rpc_url.as_str(), eth_call)
+                        .await
+                {
+                    value.rpc_url = Some(rpc_url_candidate.clone());
+                    if value.error.is_none() {
+                        return Some(value);
+                    } else {
+                        last_value = Some(value);
+                    }
+                }
             }
-        };
-        let mut value: ResponseEthCall = http_post_json(rpc_url.as_str(), eth_call).await?;
-        value.rpc_url = Some(rpc_url_candidate.to_string());
-        if value.error.is_none() {
-            return Ok(Some(value));
-        } else {
-            last_value = Some(value);
         }
+        last_value
     }
-    Ok(last_value)
 }
 
-fn process_results(results: &FetchedDataForFeed) -> Result<Payload> {
+fn process_results(results: &FetchedDataHashMap, resourses: &[FeedConfig]) -> Result<Payload> {
```

Comment:
NIT: It would be nice to change the name of the function to the kind of processing you are doing.